### PR TITLE
Add flag to prune PASSED subtests in junit xml and have top level tests

### PR DESCRIPTION
**Pull request by Rajalakshmi-Girish**
The CI which runs k8s unit tests has a huge no.of rows while it uses https://github.com/kubernetes/kubernetes/blob/master/hack/make-rules/test.sh and generates junit xml file in artifacts.

This large no. of rows hinders the display of complete data in test grids

Ref: https://github.com/kubernetes/test-infra/issues/24779
https://github.com/kubernetes/test-infra/issues/23425

As the purpose of the test grid in certain cases would be to analyze the failing/flaking tests, it shall be useful to have an option in `prune-junit-xml` to prune PASSED test cases and display only failed test cases.

### CODE REVIEW

**Comment by Rajalakshmi-Girish** [comment]
kind/feature

**Comment by liggitt** [comment]
is the expectation that all/most testgrids would use this option by default? losing info about green runs will make it hard to tell the difference between a test running and being skipped, as well as losing run-time information testgrid surfaces as graphs of times

**Comment by BenTheElder** [comment]
Can we collapse to e.g. package level results instead?

**Comment by liggitt** [comment]
> Can we collapse to e.g. package level results instead?

or even package+top-level test ... a lot of the bulk comes from tests with ~hundreds of programmatically generated subtests

**Comment by dims** [comment]
/release-note-none

cc @aojea @pohly

File: cmd/prune-junit-xml/prunexml.go
Comment (pohly) [inline cmd/prune-junit-xml/prunexml.go:99]:
```suggestion
			// The top level testcase element in a JUnit xml file does not have the / character.
```

**Review by pohly — CHANGES_REQUESTED** [review]
This looks good, thanks for taking care of this.

The code isn't that complex either, but can you perhaps add unit tests for it anyway?

File: cmd/prune-junit-xml/prunexml.go
Comment (BenTheElder) [inline cmd/prune-junit-xml/prunexml.go:93]: do we not show a top-level name for failed tests?

we can maybe leave this for later, but I think ideally for testgrid you want to have the same set of top-level identifiers across runs and if we need more granular results we make them nested under the top level names cc @michelle192837

**Comment by BenTheElder** [comment]
one comment, otherwise +1 to @pohly comments

File: cmd/prune-junit-xml/prunexml.go
Comment (michelle192837) [reply cmd/prune-junit-xml/prunexml.go:93]: Consistency on row names is usually better, yeah. If only failing results are reported, those failing results will create a row (which is the expensive part for storage/display), and weird things might happen with alerting and the like (because something that runs intermittently and fails with no examples of it passing looks like a consistent failure). I think pruning all subtests might be better in that case.

As an additional possibility to trimming all subtests:

If you want to include some more info, I think it'd be possible to report a flaky result...I believe you'd want to report the pass/fail status for each subtest as a common higher-level name, and TestGrid will merge them when processing into one cell. (This is worth testing out somewhere first though).

e.g. iirc imagine you have
- `testcase classname="k8s.io/kubernetes/test/my-stuff" name="Test1"` (pass)
- `testcase classname="k8s.io/kubernetes/test/my-stuff" name="Test2"` (fail)
- `testcase classname="k8s.io/kubernetes/test/my-stuff" name="Test3"` (pass)

Report these as `testcase classname="k8s.io/kubernetes/test/my-stuff" name="Test"` instead 3 times, 2 passing and 1 failing, and TestGrid should merge it into a single cell with a purple 'flaky' status and "2/3 runs passed".

Comment (pohly) [reply cmd/prune-junit-xml/prunexml.go:93]: I'm not sure whether I like reporting "k8s.io/kubernetes/test/my-stuff" as flaky. That is potentially misleading when the "Test2" fails reliably, i.e. there is no flake.

Comment (pohly) [reply cmd/prune-junit-xml/prunexml.go:93]: > do we not show a top-level name for failed tests?

If a sub-test failed, all parent tests are also reported as failed, i.e. won't get pruned.

Comment (BenTheElder) [reply cmd/prune-junit-xml/prunexml.go:93]: > If a sub-test failed, all parent tests are also reported as failed, i.e. won't get pruned.

Right, but currently I only see sub-test entries and no top-level entry https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit-1-27/1643306358801436672

Comment (Rajalakshmi-Girish) [reply cmd/prune-junit-xml/prunexml.go:93]: > > do we not show a top-level name for failed tests?
> 
> If a sub-test failed, all parent tests are also reported as failed, i.e. won't get pruned.

@pohly Yes, you are right! The parent tests are reporting failure when a sub-test failed. Hence all the subtests can be pruned for test grid to have the same set of rows for all runs as @BenTheElder and @michelle192837 mentioned.
I shall submit a patch to remove the condition to include failed sub-tests

@BenTheElder Below image depicts how top-level test has failed status when sub test fails:
image

Comment (Rajalakshmi-Girish) [reply cmd/prune-junit-xml/prunexml.go:93]: > > If a sub-test failed, all parent tests are also reported as failed, i.e. won't get pruned.
> 
> Right, but currently I only see sub-test entries and no top-level entry https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit-1-27/1643306358801436672
> 
> Edit: I see we're still doing package + test name, just not golang subtests.
> 
> I think we should be considering package name as the top level and nesting under that, there's already a ridiculous amount of packages and those are more stable top level identifiers

@BenTheElder In the link that you mentioned, https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit-1-27/1643306358801436672

I see both top-level and sub-test entries.
image

Pruning these subtest rows and having only top-level tests would condense the no of rows from ~39k to ~9k.

If a Top-level test indicates failure, we may get further info on what subtest failed by looking at the Build Log, which is generated from job's stdout file and hence has all the details.

Comment (pohly) [reply cmd/prune-junit-xml/prunexml.go]: > I shall submit a patch to remove the condition to include failed sub-tests

Actually, I prefer the current approach of showing exactly which sub-tests have failed also in testgrid.

If I understand @michelle192837 correctly, it will be a bit odd to see such a flaky sub-test in columns where it failed and not in those were it succeeded, but IMHO that still gives a better impression of what's going on than hiding sub-tests entirely.

Comment (Rajalakshmi-Girish) [reply cmd/prune-junit-xml/prunexml.go:93]: @pohly I have reverted the change. Can you please take a look?

File: cmd/prune-junit-xml/prunexml_test.go
Comment (pohly) [inline cmd/prune-junit-xml/prunexml_test.go:101]: I agree with @BenTheElder when he wrote:
> I think we should be considering package name as the top level and nesting under that, there's already a ridiculous amount of packages and those are more stable top level identifiers

In this example, the output then would have only `<testcase classname="k8s.io/kubernetes/cluster/gce"  name="cos"` and not mention `TestServerOverride` and `TestStorageOptions`.

The time of that artificial `testcase` should be the sum of the `testcases` that it replaces.

**Review by pohly — CHANGES_REQUESTED** [review]

File: cmd/prune-junit-xml/prunexml_test.go
Comment (Rajalakshmi-Girish) [reply cmd/prune-junit-xml/prunexml_test.go:101]: @pohly a part from package name being the top level identifiers, Ben also mentioned having other tests nested under that. Does that mean different from pruning?
We already have the option of "Group by Hierarchy Pattern", when the regular expression "\w+" is given it gets nested as below.
image

But as the data remains the same, it doesn't reduce the number of rows and hence the difficulty in loading complete data still exists.

Hope as you mentioned above, pruning all Testcase rows and creating artificial tests with
`name` entity of testcase as package name
`time` = sum of testcase durations and
`failure` message=concatinating the messages from subtests would suffice.

File: cmd/prune-junit-xml/prunexml.go
Comment (michelle192837) [reply cmd/prune-junit-xml/prunexml.go:93]: Yeah to be clear, it's odd and may mess with any alerting settings you have, but as long as that's understood you're free to do so. (I just want to be very clear because this kind of thing, only showing failures and not passes, has led to a _lot_ of confusion before. It may be worth adding a note about this behavior in the description of the tab, e.g. `description` in https://github.com/GoogleCloudPlatform/testgrid/blob/master/config.md#prow-job-configuration or https://github.com/GoogleCloudPlatform/testgrid/blob/master/config.md#tab-descriptions).

File: cmd/prune-junit-xml/prunexml_test.go
Comment (BenTheElder) [reply cmd/prune-junit-xml/prunexml_test.go:101]: > We already have the option of "Group by Hierarchy Pattern", when the regular expression "\w+" is given it gets nested as below.

TIL. @michelle192837 can we make that a default for certain pages?

**Comment by BenTheElder** [comment]
aside: this will be merging *after* kubernetes release next week due to the release policies, but we could cherry-pick it back potentially https://k8s.dev/release

File: cmd/prune-junit-xml/prunexml_test.go
Comment (Rajalakshmi-Girish) [reply cmd/prune-junit-xml/prunexml_test.go:101]: IMO, even if this filter is set by default, it would not reduce the amount of data/number of rows. Thus the problem of the page not loading all columns wouldn't be solved. @michelle192837 Please correct me if I am wrong.

@BenTheElder Please let me know if you are fine with pruning the data and having package names displayed as top-level identifiers and have consistent rows for all the runs.

Also, I would like to mention, as this is going to be an additional flag to `prune-junit-xml` tool, only the users who wish to use can include it in their CI.

Comment (Rajalakshmi-Girish) [reply cmd/prune-junit-xml/prunexml_test.go:101]: @BenTheElder Please confirm if you are fine with pruning the data and having package names displayed as top-level identifiers and have consistent rows for all the runs.

So this would create artificial tests in the junit_xml file with
`name` entity of testcase as package name
`time` = sum of testcase durations and
`failure message`=concatinating the messages from subtests

Comment (BenTheElder) [reply cmd/prune-junit-xml/prunexml_test.go:101]: Sorry I was out for a week.

> IMO, even if this filter is set by default, it would not reduce the amount of data/number of rows. Thus the problem of the page not loading all columns wouldn't be solved.

Yes, but it very nice for the humans!

I'm fine with either approach, cc @liggitt @pohly

**Comment by Rajalakshmi-Girish** [comment]
@liggitt @pohly Can you please have a look at this?

**Review by pohly — COMMENTED** [review]
The change itself looks good to me.

However, should this PR also enable the new feature? Right now it doesn't, because default is `false` and the command line flag is not getting added to the test scripts which invoke `prune-junit`.

I think this PR should enable pruning. The simplest way to do that would be to change the flag default. Only Kubernetes uses it, so we don't need to worry about breaking someone other than ourselves.

**Review by BenTheElder — COMMENTED** [review]
/lgtm
+1 to flip the default for prune-tests to true though

**Comment by Rajalakshmi-Girish** [comment]
Messed up while rebasing and resolving conflicts :( 
Please see https://github.com/kubernetes/kubernetes/pull/118487

**Comment by Rajalakshmi-Girish** [comment]
@pohly ^^

File: cmd/prune-junit-xml/prunexml.go
Comment (pohly) [inline cmd/prune-junit-xml/prunexml.go:32]: It's on by default now, as suggested in https://github.com/kubernetes/kubernetes/pull/117029#pullrequestreview-1444360343

**Review by pohly — APPROVED** [review]
/lgtm

**Comment by pohly** [comment]
Looks like we need one more approval.

/assign @dims

**Comment by dims** [comment]
/approve

thanks @Rajalakshmi-Girish

**Pull request merged**
